bot/utils/scheduling.py

```diff
@@ -1,81 +1,126 @@
-    async def _scheduled_task(self, task_object: t.Any) -> None:
-        """
-        A coroutine which handles the scheduling.
+        self._log = logging.getLogger(f"{__name__}.{name}")
+        self._scheduled_tasks: t.Dict[t.Hashable, asyncio.Task] = {}
 
-        This is added to the scheduled tasks, and should wait the task duration, execute the desired
-        code, then clean up the task.
+    def __contains__(self, task_id: t.Hashable) -> bool:
+        """Return True if a task with the given `task_id` is currently scheduled."""
+        return task_id in self._scheduled_tasks
 
-        For example, in Reminders this will wait for the reminder duration, send the reminder,
-        then make a site API request to delete the reminder from the database.
+    def schedule(self, task_id: t.Hashable, coroutine: t.Coroutine) -> None:
         """
+        Schedule the execution of a `coroutine`.
 
-    def schedule_task(self, task_id: t.Hashable, task_data: t.Any) -> None:
+        If a task with `task_id` already exists, close `coroutine` instead of scheduling it. This
+        prevents unawaited coroutine warnings. Don't pass a coroutine that'll be re-used elsewhere.
         """
-        Schedules a task.
+        self._log.trace(f"Scheduling task #{task_id}...")
 
-        `task_data` is passed to the `Scheduler._scheduled_task()` coroutine.
-        """
-        log.trace(f"{self.cog_name}: scheduling task #{task_id}...")
+        msg = f"Cannot schedule an already started coroutine for #{task_id}"
+        assert inspect.getcoroutinestate(coroutine) == "CORO_CREATED", msg
 
         if task_id in self._scheduled_tasks:
-            log.debug(
-                f"{self.cog_name}: did not schedule task #{task_id}; task was already scheduled."
-            )
+            self._log.debug(f"Did not schedule task #{task_id}; task was already scheduled.")
+            coroutine.close()
             return
 
-        task = asyncio.create_task(self._scheduled_task(task_data))
+        task = asyncio.create_task(coroutine, name=f"{self.name}_{task_id}")
         task.add_done_callback(partial(self._task_done_callback, task_id))
 
         self._scheduled_tasks[task_id] = task
-        log.debug(f"{self.cog_name}: scheduled task #{task_id} {id(task)}.")
+        self._log.debug(f"Scheduled task #{task_id} {id(task)}.")
+
+    def schedule_at(self, time: datetime, task_id: t.Hashable, coroutine: t.Coroutine) -> None:
+        """
+        Schedule `coroutine` to be executed at the given naïve UTC `time`.
+
+        If `time` is in the past, schedule `coroutine` immediately.
+
+        If a task with `task_id` already exists, close `coroutine` instead of scheduling it. This
+        prevents unawaited coroutine warnings. Don't pass a coroutine that'll be re-used elsewhere.
+        """
+        delay = (time - datetime.utcnow()).total_seconds()
+        if delay > 0:
+            coroutine = self._await_later(delay, task_id, coroutine)
```

Comment:
Not at all necessary and functionally equivalent, but this would be a good reason to use the walrus operator:

```suggestion
        if (delay := (time - datetime.utcnow()).total_seconds()) > 0:
            coroutine = self._await_later(delay, task_id, coroutine)
```

I find that it's a minor improvement to readability by clearly communicating to the reader that the variable is only used within the scope of a statement, and makes the code a bit less repetitive.

Reply:
I disagree because it makes the expression far too complicated. I wouldn't mind it if the expression started out simpler, but it already had a set of parenthesis and two function calls which just turns into a blob of parenthesis at a glance.

Reply:
Yeah, I can see that it could exceed the threshold of "reasonable amount of parenthesis". Either way, it's fairly subjective and not a huge deal, which is why I approved of the PR without it.

Reply:
I prefer it the way it is - I don't think the walrus adds any readability here, rather the opposite.